File: lib/isorun/renderer.rb
# frozen_string_literal: true

module Isorun
  class Renderer # rubocop:disable Lint/EmptyClass
    # @!method render(bundle_path, proc)
    #   Renders the JavaScript application, takes an optional callback block to
    #   control aspects of the render process (e.g. fetching data).
    #
    #   @example Render and intercept network requests to the GraphQL API
    #     renderer = Renderer.new
    #     renderer.render(bundle_path) do |action, *args|
    #       if action == "fetch"
    #         execute_graphql(*args)
    #       end
    #     end
    #
    # @param [String] bundle_path
    # @param [Proc] block
    # @return [String] The rendered result as String
    # @!visibility private
  end
end

File: ext/isorun/src/isorun/renderer.rs
use crate::isorun::js_runtime::JS_RUNTIME;
use magnus::block::Proc;
use magnus::Error;

#[magnus::wrap(class = "Isorun::Renderer")]
pub(crate) struct Renderer;

impl Renderer {
    pub(crate) fn new() -> Renderer {
        Renderer
    }

    pub(crate) fn render(
        &self,
        bundle_path: String,
        block: Proc,
    ) -> Result<String, Error> {
        JS_RUNTIME
            .with(|js_runtime| js_runtime.render(bundle_path.as_str(), block))
    }
}
